# HG changeset patch
# User Nir Krakauer <nkrakauer@ccny.cuny.edu>
# Date 1417811766 28800
#      Fri Dec 05 12:36:06 2014 -0800
# Branch stable
# Node ID 7d272300a880eb0dfa33fc3ea66eef922c7f7891
# Parent  ef649fe230f569f234d9d04646f4f94e4b527218
datevec.m: Fix negative number outputs with integer (non-float) inputs (bug #43753).

* datevec.m: Use double() to cast input before doing further calculations.
Add BIST tests to check correct behavior.

diff --git a/scripts/time/datevec.m b/scripts/time/datevec.m
--- a/scripts/time/datevec.m
+++ b/scripts/time/datevec.m
@@ -144,17 +144,17 @@ function [y, m, d, h, mi, s] = datevec (
       endfor
     endif
 
   else   # datenum input
 
     date = date(:);
 
     ## Move day 0 from midnight -0001-12-31 to midnight 0000-3-1
-    z = floor (date) - 60;
+    z = double (floor (date) - 60);
     ## Calculate number of centuries; K1 = 0.25 is to avoid rounding problems.
     a = floor ((z - 0.25) / 36524.25);
     ## Days within century; K2 = 0.25 is to avoid rounding problems.
     b = z - 0.25 + a - floor (a / 4);
     ## Calculate the year (year starts on March 1).
     y = floor (b / 365.25);
     ## Calculate day in year.
     c = fix (b - floor (365.25 * y)) + 1;
@@ -309,9 +309,11 @@ endfunction
 %% Test millisecond format FFF
 %!assert (datevec ("15:38:21.25", "HH:MM:SS.FFF"), [yr,1,1,15,38,21.025])
 
 # Other tests
 %!assert (datenum (datevec ([-1e4:1e4])), [-1e4:1e4]');
 %!test
 %! t = linspace (-2e5, 2e5, 10993);
 %! assert (all (abs (datenum (datevec (t)) - t') < 1e-5));
+%!assert (double (datevec (int64 (datenum ([2014 6 1])))), datevec (datenum ([2014 6 1])))
+%!assert (double (datevec (int64 (datenum ([2014 6 18])))), datevec (datenum ([2014 6 18])))
 
